roaring/src/bitmap/iter.rs

```diff
@@ -294,7 +316,337 @@ impl RoaringBitmap {
                 count += 1;
             }
         }
-
         Ok(count)
     }
 }
+
+mod iter_inner {
+    use crate::bitmap::container::Container;
+    use crate::bitmap::{container, util, IntoIter, Iter};
+    use core::slice;
+
+    pub(super) trait IterInternal<'a> {
+        type Container: IntoIterator<IntoIter = container::Iter<'a>, Item = u32> + AsRef<Container>;
+
+        type ContainerIterator: Iterator<Item = Self::Container> + DoubleEndedIterator;
+        type IntoContainerIterator: IntoIterator<IntoIter = Self::ContainerIterator>;
+
+        fn pop_container_front(&mut self) -> Option<Self::Container>;
+        fn pop_container_back(&mut self) -> Option<Self::Container>;
+
+        fn drain_containers_until(&mut self, index: usize);
+
+        fn clear_containers(&mut self);
+
+        fn find_container(&self, key: u16) -> Option<usize>;
+
+        fn iter_front_mut(&mut self) -> &mut container::Iter<'a>;
+        fn iter_back_mut(&mut self) -> &mut Option<container::Iter<'a>>;
+
+        fn dec_size_hint(&mut self, n: u64);
+        fn set_size_hint(&mut self, n: u64);
+
+        fn empty_inner_iter() -> container::Iter<'static>;
+
+        fn decompose(
+            self,
+        ) -> (container::Iter<'a>, Option<container::Iter<'a>>, Self::IntoContainerIterator);
+
+        #[inline]
+        fn next_inner(&mut self) -> Option<u32> {
+            self.dec_size_hint(1);
+            loop {
+                if let item @ Some(_) = self.iter_front_mut().next() {
+                    return item;
+                }
+                if self.advance_container().is_some() {
+                    continue;
+                } else {
+                    return None;
+                }
+            }
+        }
+
+        #[inline]
+        fn next_back_inner(&mut self) -> Option<u32> {
+            self.dec_size_hint(1);
+            loop {
+                if let Some(iter) = self.iter_back_mut() {
+                    if let item @ Some(_) = iter.next_back() {
+                        return item;
+                    }
+                }
+                if self.advance_container_back().is_some() {
+                    continue;
+                } else {
+                    return None;
+                }
+            }
+        }
+
+        fn advance_to_inner(&mut self, n: u32) {
+            fn advance_iter(iter: &mut container::Iter<'_>, n: u32) -> u64 {
+                let mut items_skipped = 0;
+                while let Some(next) = iter.peek() {
+                    if next < n {
+                        iter.next();
+                        items_skipped += 1;
+                    } else {
+                        return items_skipped;
+                    }
+                }
+                items_skipped
+            }
```

Comment:
Both array and bitmap container iterators should be able to do this more efficiently, we should be able to implement an `store::Iter::advance_to(&mut self, n: u16)` function that can do better than `O(n)`:

For an array container, we can binary search for `n`, and use `Iterator::nth` to efficiently skip ahead to just before the element. For a bitmap container, we should be able to just directly advance `key`, and mask the value (with some edge cases where `key` is unchanged and we need to mask the current value, or `key` becomes equal to `key_back`, and we need to steal value_back).